### PR TITLE
add metrics to consensus

### PR DESCRIPTION
**Pull request by pfi79**
#### wal

- wal_count_of_files

#### requestPool

- pool_count_of_elements
- pool_count_of_fail_add_request
- pool_count_leader_forward_request
- pool_count_timeout_two_step
- pool_count_of_delete_request
- pool_count_of_elements_all
- pool_latency_of_elements

#### blacklist

- blacklist_count
- node_id_in_blacklist

#### consensus

- consensus_reconfig
- consensus_latency_sync

#### view

- view_number
- view_leader_id
- view_phase
- view_proposal_sequence
- view_count_txs_in_batch
- view_decisions
- view_count_batch_all
- view_count_txs_all
- view_size_batch
- view_latency_batch_processing
- view_latency_batch_save

#### viewchanger

- viewchange_current_view
- viewchange_next_view
- viewchange_real_view

### CODE REVIEW

File: internal/bft/metrics.go [hunk 1]
```diff
@@ -0,0 +1,340 @@
+package bft
+
+import metrics "github.com/SmartBFT-Go/consensus/pkg/api"
+
+var countOfRequestPoolOpts = metrics.GaugeOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "pool_count_of_elements",
+	Help:         "Count of request elements in pool for this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "Number of elements in the consensus request pool.",
```

File: internal/bft/metrics.go [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var countOfFailAddRequestToPoolOpts = metrics.CounterOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "pool_count_of_fail_add_request",
+	Help:         "Count of fail add request in pool for this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "Number of requests pool insertion failure.",
```

File: internal/bft/metrics.go [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+// ForwardTimeout
+var countOfLeaderForwardRequestOpts = metrics.CounterOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "pool_count_leader_forward_request",
+	Help:         "Count leader forward request on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "Number of requests forwarded to the leader.",
```

File: internal/bft/metrics.go [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var countTimeoutTwoStepOpts = metrics.CounterOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "pool_count_timeout_two_step",
+	Help:         "Count of timeout two step on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "Number of times requests reached second timeout.",
```

File: internal/bft/metrics.go [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var countOfDeleteRequestPoolOpts = metrics.CounterOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "pool_count_of_delete_request",
+	Help:         "Count of request delete from pool on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "Number of elements removed from the request pool.",
```

File: internal/bft/metrics.go [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var countOfRequestPoolAllOpts = metrics.CounterOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "pool_count_of_elements_all",
+	Help:         "All of request elements in pool for this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "Total amount of elements in the request pool.",
```

File: internal/bft/metrics.go [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var latencyOfRequestPoolOpts = metrics.HistogramOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "pool_latency_of_elements",
+	Help:         "Latency of request elements in pool for this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "The average request processing time, time request resides in the pool.",
```

File: internal/bft/metrics.go [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+}
+
+var nodesInBlackListOpts = metrics.GaugeOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "node_id_in_blacklist",
+	Help:         "Node ID in blacklist on this channel.",
+	LabelNames:   []string{"channel", "id"},
+	StatsdFormat: "%{#fqname}.%{channel}.%{id}",
+}
```
Comment (C0rWin) [inline internal/bft/metrics.go]: Does it have a list of IDs or only one ID? There might be more than one blaklisted node.

File: internal/bft/metrics.go [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+
+type MetricsBlacklist struct {
```
Comment (C0rWin) [inline internal/bft/metrics.go]: Comment (go-doc)?

File: internal/bft/metrics.go [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "pool_latency_of_elements",
+	Help:         "Latency of request elements in pool for this channel.",
+	Buckets:      []float64{0.005, 0.01, 0.015, 0.05, 0.1, 1, 10},
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+type MetricsRequestPool struct {
```
Comment (C0rWin) [inline internal/bft/metrics.go]: go doc?

File: internal/bft/metrics.go [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+}
+
+type MetricsBlacklist struct {
+	CountBlackList   metrics.Gauge
+	NodesInBlackList metrics.Gauge
+
+	labels map[string]string
+}
+
+func NewMetricsBlacklist(p *metrics.CustomerProvider) *MetricsBlacklist {
```
Comment (C0rWin) [inline internal/bft/metrics.go]: go doc?

File: internal/bft/metrics.go [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+
+func (m *MetricsBlacklist) Label(name string) string {
+	return m.labels[name]
+}
+
+var consensusReconfigOpts = metrics.CounterOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "consensus_reconfig",
+	Help:         "count consensus reconfig on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "Number of reconfiguration requests.",
```

File: internal/bft/metrics.go [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var latencySyncOpts = metrics.HistogramOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "consensus_latency_sync",
+	Help:         "latency sync on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "An average time it takes to sync node.",
```

File: internal/bft/metrics.go [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+	Buckets:      []float64{0.005, 0.01, 0.015, 0.05, 0.1, 1, 10},
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+type MetricsConcensus struct {
```
Comment (C0rWin) [inline internal/bft/metrics.go]: go-doc?

File: internal/bft/metrics.go [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+	CountConsensusReconfig metrics.Counter
+	LatencySync            metrics.Histogram
+}
+
+func NewMetricsConcensus(p *metrics.CustomerProvider) *MetricsConcensus {
```
Comment (C0rWin) [inline internal/bft/metrics.go]: go doc?

File: internal/bft/metrics.go [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+		CountConsensusReconfig: p.NewCounter(consensusReconfigOpts).With("channel", ch),
+		LatencySync:            p.NewHistogram(latencySyncOpts).With("channel", ch),
+	}
+}
+
+var viewNumberOpts = metrics.GaugeOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "view_number",
+	Help:         "number of view on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "The View number value",
```

File: internal/bft/metrics.go [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var leaderIDOpts = metrics.GaugeOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "view_leader_id",
+	Help:         "leader id on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "The leader id.",
```

File: internal/bft/metrics.go [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var proposalSequenceOpts = metrics.GaugeOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "view_proposal_sequence",
+	Help:         "proposal sequence view on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "The sequence number within current view.",
```

File: internal/bft/metrics.go [hunk 19]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var decisionsInViewOpts = metrics.GaugeOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "view_decisions",
+	Help:         "decisions in view on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "The number of decisions in the current view.",
```

File: internal/bft/metrics.go [hunk 20]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var phaseOpts = metrics.GaugeOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "view_phase",
+	Help:         "phase in view on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "Current consensus phase.",
```

File: internal/bft/metrics.go [hunk 21]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var countTxsInBatchOpts = metrics.GaugeOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "view_count_txs_in_batch",
+	Help:         "count txs in batch on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "The number of transactions per batch.",
```

File: internal/bft/metrics.go [hunk 22]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var countBatchAllOpts = metrics.CounterOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "view_count_batch_all",
+	Help:         "count processed batch on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "Amount of batched processed.",
```

File: internal/bft/metrics.go [hunk 23]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var countTxsAllOpts = metrics.CounterOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "view_count_txs_all",
+	Help:         "count txs all on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "Total amount of transactions.",
```

File: internal/bft/metrics.go [hunk 24]
```diff
@@ -0,0 +1,340 @@
...
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "view_latency_batch_save",
+	Help:         "latency batch save on this channel.",
+	Buckets:      []float64{0.005, 0.01, 0.015, 0.05, 0.1, 1, 10},
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+type MetricsView struct {
```
Comment (C0rWin) [inline internal/bft/metrics.go]: go doc?

File: internal/bft/metrics.go [hunk 25]
```diff
@@ -0,0 +1,340 @@
...
+	Help:         "count txs all on this channel.",
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var sizeOfBatchOpts = metrics.CounterOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "view_size_batch",
+	Help:         "size of batch on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "An average batch size.",
```

File: internal/bft/metrics.go [hunk 26]
```diff
@@ -0,0 +1,340 @@
...
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var latencyBatchProcessingOpts = metrics.HistogramOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "view_latency_batch_processing",
+	Help:         "latency batch processing on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "Amount of time it take to process batch.",
```

File: internal/bft/metrics.go [hunk 27]
```diff
@@ -0,0 +1,340 @@
...
+	Buckets:      []float64{0.005, 0.01, 0.015, 0.05, 0.1, 1, 10},
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+var latencyBatchSaveOpts = metrics.HistogramOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "view_latency_batch_save",
+	Help:         "latency batch save on this channel.",
```
Comment (C0rWin) [inline internal/bft/metrics.go]:
```suggestion
	Help:         "An average time it takes to persist batch.",
```

File: internal/bft/metrics.go [hunk 28]
```diff
@@ -0,0 +1,340 @@
...
+	Phase                  metrics.Gauge
+	CountTxsInBatch        metrics.Gauge
+	CountBatchAll          metrics.Counter
+	CountTxsAll            metrics.Counter
+	SizeOfBatch            metrics.Counter
+	LatencyBatchProcessing metrics.Histogram
+	LatencyBatchSave       metrics.Histogram
+}
+
+func NewMetricsView(p *metrics.CustomerProvider) *MetricsView {
```
Comment (C0rWin) [inline internal/bft/metrics.go]: go doc?

File: internal/bft/metrics.go [hunk 29]
```diff
@@ -0,0 +1,340 @@
...
+	ch := p.Labels["channel"]
```
Comment (C0rWin) [inline internal/bft/metrics.go]: This leaks the notion of the channel into consensus. This is supposed to be hidden and property of the metrics provider side.

File: internal/bft/metrics.go [hunk 30]
```diff
@@ -0,0 +1,340 @@
...
+var realViewOpts = metrics.GaugeOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "viewchange_real_view",
+	Help:         "real view of viewchange on this channel.",
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+type MetricsViewChange struct {
```
Comment (C0rWin) [inline internal/bft/metrics.go]: go doc?

File: internal/bft/metrics.go [hunk 31]
```diff
@@ -0,0 +1,340 @@
...
+	CurrentView metrics.Gauge
+	NextView    metrics.Gauge
+	RealView    metrics.Gauge
+}
+
+func NewMetricsViewChange(p *metrics.CustomerProvider) *MetricsViewChange {
```
Comment (C0rWin) [inline internal/bft/metrics.go]: go doc?

File: internal/bft/metrics.go [hunk 32]
```diff
@@ -0,0 +1,340 @@
...
+	ch := p.Labels["channel"]
```
Comment (C0rWin) [inline internal/bft/metrics.go]: Same as the comment above about leaking abstraction. This should be hidden within the implementation and metrics provider side.

File: pkg/api/metrics.go [hunk 33]
```diff
@@ -0,0 +1,194 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package api
+
+type CustomerProvider struct {
```
Comment (C0rWin) [inline pkg/api/metrics.go]: go doc

File: pkg/api/metrics.go [hunk 34]
```diff
@@ -0,0 +1,194 @@
...
+	Provider
+	Labels map[string]string
+}
+
+func NewCustomerProvider(mp Provider, labelValues ...string) *CustomerProvider {
```
Comment (C0rWin) [inline pkg/api/metrics.go]: go doc

File: pkg/wal/metrics.go [hunk 35]
```diff
@@ -0,0 +1,24 @@
...
+var countOfFilesOpts = metrics.GaugeOpts{
+	Namespace:    "consensus",
+	Subsystem:    "bft",
+	Name:         "wal_count_of_files",
+	Help:         "Count of wal-files in this channel.",
+	LabelNames:   []string{"channel"},
+	StatsdFormat: "%{#fqname}.%{channel}",
+}
+
+type Metrics struct {
```
Comment (C0rWin) [inline pkg/wal/metrics.go]: go doc

File: pkg/wal/metrics.go [hunk 36]
```diff
@@ -0,0 +1,24 @@
...
+	CountOfFiles metrics.Gauge
+}
+
+func NewMetrics(p *metrics.CustomerProvider) *Metrics {
```
Comment (C0rWin) [inline pkg/wal/metrics.go]: go doc

**Review by C0rWin — CHANGES_REQUESTED** [review]
I think it will be much easier if you can split this change into several:

1. One to add metric interfaces
2. Separate commit for each group of metrics


Also, no unit tests are provided to ensure metrics are working and reporting as expected. Could you please take care to add them?

File: internal/bft/metrics.go (hunk 27 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 26 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 25 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 23 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 1 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 2 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 3 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 4 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 5 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 6 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 7 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 12 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 13 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 16 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 17 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 18 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 19 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 20 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 21 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 22 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: pkg/wal/metrics.go (hunk 36 above)
Comment (pfi79) [reply pkg/wal/metrics.go]: fix

File: pkg/wal/metrics.go (hunk 35 above)
Comment (pfi79) [reply pkg/wal/metrics.go]: fix

File: pkg/api/metrics.go (hunk 34 above)
Comment (pfi79) [reply pkg/api/metrics.go]: fix

File: pkg/api/metrics.go (hunk 33 above)
Comment (pfi79) [reply pkg/api/metrics.go]: fix

File: internal/bft/metrics.go (hunk 9 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 11 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 10 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 14 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 15 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 31 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 30 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 28 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 24 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 29 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/metrics.go (hunk 32 above)
Comment (pfi79) [reply internal/bft/metrics.go]: fix

File: internal/bft/util.go [hunk 37]
```diff
@@ -467,6 +479,23 @@ func (bl blacklist) computeUpdate() []uint64 {
...
+					bl.metricsBlacklist.LabelsForWith(nameBlackListNodeID, strconv.FormatUint(bl.nodes[n], 10))...,
+				).Set(1)
+				continue Outer
+			}
+		}
+		bl.metricsBlacklist.NodesInBlackList.With(
+			bl.metricsBlacklist.LabelsForWith(nameBlackListNodeID, strconv.FormatUint(bl.nodes[n], 10))...,
+		).Set(0)
+	}
+
```
Comment (C0rWin) [inline internal/bft/util.go]: I think you can make it simpler:

```
newBlacklistMap := make(map[uint64]bool, len(newBlacklist))
for _, node := range newBlacklist {
    newBlacklistMap[node] = true
}

for _, node := range bl.nodes {
    var value float64
    if newBlacklistMap[node] {
        value = 1
    } else {
        value = 0
    }

    bl.metricsBlacklist.NodesInBlackList.With(
        bl.metricsBlacklist.LabelsForWith(nameBlackListNodeID, strconv.FormatUint(node, 10))...,
    ).Set(value)
}

```

Comment (C0rWin) [reply internal/bft/util.go]: or

```
newBlacklistMap := make(map[uint64]bool, len(newBlacklist))
for _, node := range newBlacklist {
    newBlacklistMap[node] = true
}

for _, node := range bl.nodes {
    inBlacklist := newBlacklistMap[node]

    bl.metricsBlacklist.NodesInBlackList.With(
        bl.metricsBlacklist.LabelsForWith(nameBlackListNodeID, strconv.FormatUint(node, 10))...,
    ).Set(btoi(inBlacklist))
}

func btoi(b bool) float64 {
    if b {
        return 1
    }
    return 0
}

```

Comment (pfi79) [reply internal/bft/util.go]: fix

File: internal/bft/requestpool.go [hunk 38]
```diff
@@ -205,6 +214,7 @@ func (rp *Pool) Submit(request []byte) error {
 	defer cancel()
 	// do not wait for a semaphore with a lock, as it will prevent draining the pool.
 	if err := rp.semaphore.Acquire(ctx, 1); err != nil {
+		rp.metrics.CountOfFailAddRequestToPool.Add(1)
```
Comment (C0rWin) [inline internal/bft/requestpool.go]: IMO, we need to distinguish between failures cause they might be symptoms of different issues.

Comment (pfi79) [reply internal/bft/requestpool.go]: I can label it.

Comment (pfi79) [reply internal/bft/requestpool.go]: added the labels

**Pull request closed without merging**